File: 이것이 취업을 위한 코딩 테스트다/그래프_이론/커리큘럼/커리큘럼.py
from collections import deque

n = int(input())
subject = [0] * (n + 1)
indegree = [0] * (n + 1)
graph = [[] for _ in range(n + 1)]

for i in range(1, n + 1):
  arr = list(map(int, input().split()))
  subject[i] = arr[0]

  for j in arr[1:-1]:
    graph[j].append(i)
    indegree[i] += 1

print(indegree)
print(graph)
queue = deque()
result = subject[:]

for i in range(n):
  if indegree[i] == 0:
    queue.append(i)

while (queue):
  now = queue.popleft()
  print(now, end=" ")

  for i in graph[now]:
    result[i] = max(result[i], subject[i] + result[now])
    indegree[i] -= 1

    if indegree[i] == 0:
      queue.append(i)

print()

for i in range(1, n + 1):
  print(result[i])

File: 이것이 취업을 위한 코딩 테스트다/그래프_이론/커리큘럼/커리큘럼.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

vector<int> graph[501];
queue<int> que;
int indegree[501];
int times[501];
int result[501];
int n;

int main() {
  cin >> n;

  for (int i = 1; i <= n; i++) {
    int time;
    cin >> time;
    times[i] = time;

    while (true) {
      int outputNode;
      cin >> outputNode;

      if (outputNode == -1) {
        break;
      }

      graph[outputNode].push_back(i);
      indegree[i] += 1;
    }
  }

  for (int i = 1; i <= n; i++) {
    result[i] = times[i];

    if (indegree[i] == 0) {
      que.push(i);
    }
  }

  while (!que.empty()) {
    int now = que.front();
    que.pop();

    for (int i = 0; i < graph[now].size(); i++) {
      int outputNode = graph[now][i];
      result[outputNode] = max(result[outputNode], result[now] + times[outputNode]);
      indegree[outputNode] -= 1;
      if (indegree[outputNode] == 0) {
        que.push(outputNode);
      }
    }
  }

  for (int i = 1; i <= n; i++) {
    cout << result[i] << " ";
  }

  return 0;
}